packages/hoppscotch-selfhost-web/src/platform/collections/desktop/sync.ts

```diff
@@ -0,0 +1,543 @@
+
+// restCollectionsMapper uses the collectionPath as the local identifier
+export const restCollectionsMapper = createMapper<string, string>()
+
+// restRequestsMapper uses the collectionPath/requestIndex as the local identifier
+export const restRequestsMapper = createMapper<string, string>()
+
+// temp implementation untill the backend implements an endpoint that accepts an entire collection
+// TODO: use importCollectionsJSON to do this
+const recursivelySyncCollections = async (
+  collection: HoppCollection,
+  collectionPath: string,
+  parentUserCollectionID?: string
+) => {
+  let parentCollectionID = parentUserCollectionID
+
+  // if parentUserCollectionID does not exist, create the collection as a root collection
+  if (!parentUserCollectionID) {
+    const res = await createRESTRootUserCollection(collection.name)
+
+    if (E.isRight(res)) {
+      parentCollectionID = res.right.createRESTRootUserCollection.id
+
+      collection.id = parentCollectionID
+      removeDuplicateRESTCollectionOrFolder(parentCollectionID, collectionPath)
+    } else {
+      parentCollectionID = undefined
+    }
+  } else {
+    // if parentUserCollectionID exists, create the collection as a child collection
+    const res = await createRESTChildUserCollection(
+      collection.name,
+      parentUserCollectionID
+    )
+
+    if (E.isRight(res)) {
+      const childCollectionId = res.right.createRESTChildUserCollection.id
+
+      collection.id = childCollectionId
+
+      removeDuplicateRESTCollectionOrFolder(
+        childCollectionId,
+        `${collectionPath}`
+      )
+    }
+  }
+
+  // create the requests
+  if (parentCollectionID) {
+    collection.requests.forEach(async (request) => {
+      const res =
+        parentCollectionID &&
+        (await createRESTUserRequest(
+          request.name,
+          JSON.stringify(request),
+          parentCollectionID
+        ))
+
+      if (res && E.isRight(res)) {
+        const requestId = res.right.createRESTUserRequest.id
+
+        request.id = requestId
+      }
+    })
+  }
+
+  // create the folders aka child collections
+  if (parentCollectionID)
+    collection.folders.forEach(async (folder, index) => {
+      recursivelySyncCollections(
+        folder,
+        `${collectionPath}/${index}`,
+        parentCollectionID
+      )
+    })
+}
+
+// TODO: generalize this
+// TODO: ask backend to send enough info on the subscription to not need this
+export const collectionReorderOrMovingOperations: {
+  sourceCollectionID: string
+  destinationCollectionID?: string
+  reorderOperation: {
+    fromPath: string
+    toPath?: string
+  }
+}[] = []
+
+type OperationStatus = "pending" | "completed"
+
+type OperationCollectionRemoved = {
+  type: "COLLECTION_REMOVED"
+  collectionBackendID: string
+  status: OperationStatus
+}
+
+export const restCollectionsOperations: Array<OperationCollectionRemoved> = []
+
+export const storeSyncDefinition: StoreSyncDefinitionOf<
```

Comment:
We'll need to port over applicable entries from the platform definition scoped under `web`. For instance, duplicate collection action isn't synced upstream in the Desktop App. Similarly, other definitions (not limited to collections) need to be evaluated and accounted for.